feat(chart): Add a chart/ package.

technosophos opened the pull request:
This package provides a chart struct, a chartfile struct, and methods to load charts either from directories or from tgz files. The testdata directory contains some variations on charts .

Also, added a very thin log abstraction layer to give errors, warnings, info, and debug messages some consistency.

technosophos committed:
Add chart file support.

Also added a log abstraction.

technosophos committed:
Add Chart tar reader, dir reader, tests.

technosophos committed:
Load chartfile into chart on chart file load.

Best commit message ever.

technosophos commented:
@adamreese If you have a second to look at this, I'd appreciate it.

vaikas-google commented on chart/chart.go:
```diff
@@ -0,0 +1,214 @@
+package chart
```
Can you add the appropriate copyright headers to these files?

vaikas-google commented on chart/chart.go:
```diff
@@ -0,0 +1,214 @@
+}
+
+func loadTar(r *tar.Reader) (*tarChart, error) {
+	td, err := ioutil.TempDir("", "chart-")
```
Is there any reason why we need to use a temp directory (and the associated management) instead of using a package for the tarball?
https://golang.org/pkg/archive/tar/

jackgr replied:
+1

See [this code](https://github.com/kubernetes/deployment-manager/blob/master/expandybird/expander/expander.go#L51) for an example, or better yet, reuse it, if possible.

technosophos replied:
The package here implements two backends for a chart: One is the uncompressed, unpackaged directory. The other is the tar (which does use `archive/tar`). Both adhere to the same interface.

For a tar file, the code dumps the data into a temp directory. There are two major reasons to do this the way it's done here:

1. Given the new format, the sizes of charts are less predictable, so purely from the memory management angle, we didn't want to keep everything in memory. (To wit, Sebastian just suggested adding SVG image files to charts.) Nor does it make a lot of sense to keep a bunch of readers open, and try to manage concurrency with those.
2. Given what we discussed about generation headers, we now perform a number of operations on directories and on files (as path-relative) in directories. We also now explicitly allow generators to create files inside of the package. All of these are more easily accomplished by extracting the archive to a temp directory and then allowing them to be modified as files on a filesystem.

Maybe I didn't make it clear what I'm trying to do with this package. This is supposed to be the standard package for working with charts. Special purpose chart handling done elsewhere (like the expandybird stuff) should be refactored to use this. The client code will be using this as well. So I'm not advocating having duplicate code. I'm advocating rewriting all of the other code to use this one implementation, which adheres to the standard we just all agreed upon.

FWIW, a follow-up PR that I am working on has the code necessary for saving a chart, and also packing a chart into a tarball. I just didn't want to add a huge PR with everything jammed in it.

jackgr replied:
Makes sense.

vaikas-google commented on chart/chartfile.go:
```diff
@@ -0,0 +1,79 @@
+// VersionOK returns true if the given version meets the constraints.
+//
+// It returns false if the version string or constraint is unparsable or if the
+// version does not meet the constraint.
```
Just curious if it would make sense to return an error here indicating what part failed for the user. If it's too low, too high, or something like that? Might be easier to surface to the user if necessary.

jackgr replied:
+1

technosophos replied:
Filed an upstream feature request to the SemVer library: https://github.com/Masterminds/semver/issues/2

technosophos replied:
Again this is part of a test fixture. I can replace the text with lorem if you want.

vaikas-google commented on chart/testdata/frobnitz/templates/wordpress.jinja.schema:
```diff
@@ -0,0 +1,69 @@
+    properties:
+      ip:
+        type: string
+        default: 10.0.253.247
```
do we want this as a default?

jackgr replied:
We have a wordpress example [here](https://github.com/kubernetes/deployment-manager/tree/master/examples/wordpress). Can we reuse it instead of building a new one?

technosophos replied:
This is just test data, and it's copied verbatim from existing templates. It's used only in testing chart parsing.

jackgr replied:
If it's copied verbatim, can we just reference it, instead?

technosophos replied:
No, because if upstream changes, it may invalidate tests spuriously.

jackgr replied:
Red flag. We should be able to import a specific version of a chart reliably.

vaikas-google commented on glide.lock:
```diff
@@ -0,0 +1,51 @@
+hash: db55a031aaa2f352fa5e9e4fda871039afb80e383a57fc77e4b35114d47cca8a
```
I'm not familiar with glide, do you check the lock file in typically?

jackgr replied:
Looks like it adds a fair amount of complexity. What does it buy us?

technosophos replied:
Yes, we definitely check in the lock file. Glide allows developers to specify version ranges in the main glide.yaml file. It then automatically generates the glide.lock file. You can then rebuild an exact image of the last known good build using `glide install`. That reads the lock file and installs exactly the referenced versions. (`glide up`, in contrast, will re-scan the code and see if it can update things and still match the version constraints specified.)

jackgr replied:
I see. That does sound useful. Also [attributed to you](http://engineeredweb.com/blog/2014/glide-go-package-management/). Nice. There's a theme here.:0)

It looks/sounds a bit like npm, which I like a lot. Was that the mental model for it?

How does it compare with godeps/vendoring?

technosophos replied:
Yeah, it's modeled after a bunch of projects, including CPAN, Gems, npm, Composer, and Mix. We did a pretty deep analysis of package managers. And I wouldn't really call it my brainchild. When I started it a few years ago, it barely resembled what it is now.

So it uses the official Go vendor/ recommendation, and caches dependencies in a vendor folder. The lock model is similar to godeps. But it solves some additional problems.

adamreese commented on chart/chart.go:
```diff
@@ -0,0 +1,214 @@
+	preTemplates string = "templates/"
+	preHooks     string = "hooks/"
+	preDocs      string = "docs/"
+)
```
Are these const used anywhere?

technosophos replied:
No, but they will be in a subsequent commit. Maybe I'll just add those functions to this commit.

I'm also wondering if I should make them public, like `DirTemplates`.

jackgr replied:
YAGNI

Let's add them if/when they're needed.

adamreese commented:
seems sane to me

technosophos committed:
Add short/long/local names according to new spec.

The url package adds short, long, and local URL parsing.

It also supports switching between short and long names.

jackgr commented on chart/testdata/frobnitz/LICENSE:
```diff
@@ -0,0 +1,202 @@
+
```
Do we need another copy of the Apache license? I don't think so. The one at the root covers the entire contents of the repository.

technosophos replied:
This is test data. It's a complete chart that the test suite uses to test whether it parses a standards-compliant chart. I can remove the text... but it's just here because it's test data.

jackgr replied:
If we don't intend to release it independently (i.e., if it's part of this code base), then we don't need a separate license for it.

jackgr commented on chart/testdata/frobnitz/templates/wordpress.jinja:
```diff
@@ -0,0 +1,72 @@
+#helm:generate dm_template
```
We have a wordpress example [here](https://github.com/kubernetes/deployment-manager/tree/master/examples/wordpress). Can we reuse it instead of building a new one?

technosophos replied:
This is test data. The idea was to have one particular isolated example that we can run tests against. Ideally, it only changes when we need to change it for test integrity. So we can't just pull in live data from somewhere else.

jackgr commented on chart/testdata/ill/LICENSE:
```diff
@@ -0,0 +1,202 @@
+
```
See the previous comment about multiple license files.

technosophos committed:
Streamline local file handling, test Windows files

jackgr commented:
Would like to see what the refactoring of the template management code (i.e., expandybird/expander/expander.go) looks like with this package in place. Can you push a WIP for that?

technosophos commented:
The previous URL package has been moved to `chart.Locator` (to distinguish it from a full URL). It's now in the PR. I also added some tests on Windows paths, following a blog from Microsoft on the topic (given that I know nothing about Windows).

technosophos committed:
Added Locator, added license blocks.

technosophos committed:
Add vendor/ to gitignore.

technosophos committed:
Refactor Chart to be a struct, with more methods.

This refactors the Chart to be a struct that can have numerous
properties and methods. Now, a chartLoader implements the
loading and closing of charts.

technosophos committed:
Add Save and Create functions.

jackgr commented:
Merging to get these changes in. Please fix the remaining issues.
